evm/v0.5/testv2/Coordinator.test.ts

```diff
@@ -0,0 +1,909 @@
+  )
+}
+
+const newServiceAgreement = async (
+  sA?: Partial<h.ServiceAgreement>,
+): Promise<h.ServiceAgreement> => {
+  return h.newServiceAgreement({
+    oracles: [roles.oracleNode.address],
+    ...partialServiceAgreement,
+    ...sA,
+  })
+}
+
+let link: Instance<LinkTokenFactory>
+let coordinator: Instance<CoordinatorFactory>
+let emptyAggregator: Instance<EmptyAggregatorFactory>
+let meanAggregator: Instance<MeanAggregatorFactory>
+let getterSetter: Instance<GetterSetterFactory>
+let oracle1: string, oracle2: string, oracle3: string
+let oracles: string[]
+
+const deployment = h.useSnapshot(provider, async () => {
+  link = await linkTokenFactory.connect(roles.defaultAccount).deploy()
+  coordinator = await coordinatorFactory
+    .connect(roles.defaultAccount)
+    .deploy(link.address)
+  emptyAggregator = await emptyAggregatorFactory
+    .connect(roles.defaultAccount)
+    .deploy()
+  meanAggregator = await meanAggregatorFactory
+    .connect(roles.defaultAccount)
+    .deploy()
+  getterSetter = await getterSetterFactory
+    .connect(roles.defaultAccount)
+    .deploy()
+})
+
+beforeAll(async () => {
+  roles = await h.initializeRolesAndPersonas(provider).then(x => x.roles)
+  oracle1 = roles.oracleNode1.address
+  oracle2 = roles.oracleNode2.address
+  oracle3 = roles.oracleNode3.address
+  oracles = [oracle1, oracle2, oracle3]
+})
+
+beforeEach(async () => {
+  await deployment()
+  link = link.connect(roles.defaultAccount)
+  coordinator = coordinator.connect(roles.defaultAccount)
+  emptyAggregator = emptyAggregator.connect(roles.defaultAccount)
+  meanAggregator = meanAggregator.connect(roles.defaultAccount)
+  getterSetter = getterSetter.connect(roles.defaultAccount)
+})
+
+describe('Coordinator', () => {
+  it('has a limited public interface', () => {
+    h.checkPublicABI(coordinatorFactory, [
+      'EXPIRY_TIME',
+      'balanceOf',
+      'cancelOracleRequest',
+      'depositFunds',
+      'fulfillOracleRequest',
+      'getId',
+      'initiateServiceAgreement',
+      'onTokenTransfer',
+      'oracleRequest',
+      'serviceAgreements',
+      'withdraw',
+      'withdrawableTokens',
+    ])
+  })
+
+  describe('#getId', () => {
+    it('matches the ID generated by the oracle off-chain', async () => {
+      const agreement = await newServiceAgreement({
+        payment: 1,
+        expiration: 2,
+        requestDigest:
+          '0x85820c5ec619a1f517ee6cfeff545ec0ca1a90206e1a38c47f016d4137e801dd',
+        aggregator: emptyAggregator.address,
+      })
+      const sAID = h.generateSAID(agreement)
+      const sAAsData = h.encodeServiceAgreement(agreement)
+      const result = await coordinator.getId(sAAsData)
+      assert.equal(result.toLowerCase(), sAID)
+    })
+  })
+
+  describe('#initiateServiceAgreement', () => {
+    describe('with valid oracle signatures', () => {
+      let serviceAgreement: h.ServiceAgreement,
+        sAID: string,
+        receipt: ethers.providers.TransactionReceipt
```

Comment:
This is actually the type of `ContractReceipt` which is a subtype of `TransactionReceipt`, it exposes the `events` property for us.